circuits/src/lib.rs

```diff
@@ -461,6 +461,16 @@ pub trait SingleProverCircuit {
     /// The prover commits to the witness and sends this commitment to the verifier, this type
     /// is the structure in which that commitment is sent
     type WitnessCommitment;
+    /// The type of a witness that has been allocated into a constraint system
+    ///
+    /// This is created by both the prover and the verifier in preparation for
+    /// applying constraints
+    type WitnessVar;
+    /// The type of a statement that has been allocated into a constraint system
+    ///
+    /// This is created by both the prover and the verifier in preparation for
+    /// applying constraints
+    type StatementVar;
```

Comment:
I don't think we need to add two more associated types here. I'd suggest a different approach that more closely aligns with the existing traits:
1. Restrict the above types as `type Witness: CommitWitness` and `type Statement: Clone + CommitPublic`
2. In `apply_constraints` we type `witness_var: <Self::Witness as CommitWitness>::VarType` and `statement_var: <Self::Statement as CommitPublic>::VarType`

This more directly expresses the relationship between statement and witness vars, and enforces a trait bound that should really exist anyways.

Reply:
I'll also make a more generic change if this is done (outside of this PR), to abstract prove/verify using your new `apply_constraints` method

Reply:
I like this, gonna go ahead and take a stab!